### PR TITLE
Distinst InstallOptions

### PR DESCRIPTION
**Pull request by mmstick**
Distinst provides a collection of possible installation options. The installer may select the corresponding option that it wishes to select, and perform an installation with that option. At the moment, the installer does not yet have the UI to support all of installation options that are currently supported by distinst.

Required by https://github.com/elementary/installer/issues/320

### CODE REVIEW

**Comment by danrabbit** [comment]
Hey @mmstick I definitely appreciate all the work you guys have done here. Is there any way we can break this up into multiple PRs or at least multiple commits? A branch this large is going to be difficult to review and kind of defeats the ability to do any kind of revision control

**Comment by mmstick** [comment]
@danrabbit I can try, but it will take quite a bit longer to accomplish.

**Comment by danrabbit** [comment]
At the very least it seems like there are quite a lot of code style issues. There's a number of times in the diff where extra spaces are introduced or spaces are removed or methods are just moved. It's kind of a messy diff :grimacing:

File: src/Widgets/DiskBar.vala [hunk 1]
```diff
@@ -24,113 +24,88 @@ public class Installer.DiskBar: Gtk.Grid {
...
-        attach (bar, 1, 1);
-
-        show_all ();
     }
 
     private void add_legend (string ppath, uint64 size, string fs, string? vg, Gtk.Popover? menu) {
         var fill_round = new FillRound ();
         fill_round.width_request = fill_round.height_request = 14;
-        fill_round.valign = Gtk.Align.CENTER;
+        fill_round.set_valign(Gtk.Align.CENTER);
```
Comment (cassidyjames) [inline src/Widgets/DiskBar.vala]: There are a lot of switching back to setter methods instead of properties. This is probably just a weird merge artifact.

File: src/Widgets/DiskBar.vala [hunk 2]
```diff
@@ -24,113 +24,88 @@ public class Installer.DiskBar: Gtk.Grid {
...
         context.add_class ("legend");
         context.add_class (fs);
 
         var format_size = GLib.format_size (size);
-
         var info = new Gtk.Label (
             (vg == null)
-                ? _("%s (%s)").printf (format_size, fs)
-                : _("%s (%s: <b>%s</b>)").printf (format_size, fs, vg)
+                ? "%s (%s)".printf (format_size, fs)
```
Comment (cassidyjames) [inline src/Widgets/DiskBar.vala]: I believe these should still be translatable.

File: src/Widgets/DiskBar.vala [hunk 3]
```diff
@@ -24,113 +24,88 @@ public class Installer.DiskBar: Gtk.Grid {
...
+                : "%s (%s: <b>%s</b>)".printf (format_size, fs, vg)
         );
         info.use_markup = true;
 
         var path = new Gtk.Label ("<b>%s</b>".printf (ppath));
-        path.halign = Gtk.Align.START;
+        path.set_halign (Gtk.Align.START);
```
Comment (cassidyjames) [inline src/Widgets/DiskBar.vala]: another unwanted change to a setter instead of property

File: src/Widgets/DiskBar.vala [hunk 4]
```diff
@@ -148,7 +123,7 @@ public class Installer.DiskBar: Gtk.Grid {
             event_box.add (widget);
             event_box.add_events (Gdk.EventMask.BUTTON_PRESS_MASK);
             event_box.button_press_event.connect (() => {
-                menu.popup ();
+                menu.popup();
```
Comment (cassidyjames) [inline src/Widgets/DiskBar.vala]: Weird merge artifact? Space before paren is proper code style.

File: src/Dialogs/HelpDialog.vala [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            resizable: false
+        );
+    }
+
+    construct {
+        var header = new Gtk.HeaderBar ();
+        header.set_show_close_button (true);
+        var header_context = header.get_style_context ();
+
+        var dual_booting = new Gtk.Label ("""<b>Create Partitions</b>
```
Comment (cassidyjames) [inline src/Dialogs/HelpDialog.vala]: I know I wrote this whole dialog, but I'm not super happy with it. It's big and untranslatable. There's probably a better way to solve this, and this is a prime candidate to be broken off into its own PR.

File: src/MainWindow.vala [hunk 6]
```diff
@@ -35,14 +35,18 @@ public class Installer.MainWindow : Gtk.Dialog {
 
     private uint64 minimum_disk_size;
 
+    private DateTime? start_date = null;
+    private DateTime? end_date = null;
```
Comment (cassidyjames) [inline src/MainWindow.vala]: For the pop-specific timer thing that I think we can remove now.

File: src/MainWindow.vala [hunk 7]
```diff
@@ -165,10 +184,12 @@ public class Installer.MainWindow : Gtk.Dialog {
...
         });
 
-        disk_view.next_step.connect (() => load_encrypt_view ());
+        disk_view.next_step.connect (() =>  load_encrypt_view ());
+
+        load_check_view ();
     }
 
-    private void load_partitioning_view () {
+    private void load_partitioning_view() {
```
Comment (cassidyjames) [inline src/MainWindow.vala]: space before parens :smile:

File: src/Widgets/DiskBar.vala (hunk 2 above)
Comment (mmstick) [reply src/Widgets/DiskBar.vala]: I don't see how, since `%s` doesn't translate.

Comment (cassidyjames) [reply src/Widgets/DiskBar.vala]: The ordering of the strings can be region-specific, as can the use of parens versus other characters.

File: src/Widgets/DiskBar.vala (hunk 4 above)
Comment (mmstick) [reply src/Widgets/DiskBar.vala]: It could be that it was fixed upstream, but not in our fork.

File: src/MainWindow.vala (hunk 7 above)
Comment (mmstick) [reply src/MainWindow.vala]: I'll be using `vala-lint` to fix all the style issues

**Comment by mmstick** [comment]
I think I have all the major issues addressed now. There's now more lines code removed in this PR than there are added. Now I just need to split this up into smaller PRs.

File: src/Config.vala.in [hunk 8]
```diff
@@ -1,7 +1,5 @@
 namespace Build {
-    public const string GETTEXT_PACKAGE = "@GETTEXT_PACKAGE@";
-    public const string LANG_LIST = "@LANG_LIST@";
-    public const string PREFERRED_LANG_LIST = "@PREFERRED_LANG_LIST@";
-    public const string SQUASHFS_PATH = "@SQUASHFS_PATH@";
-    public const string MANIFEST_REMOVE_PATH = "@MANIFEST_REMOVE_PATH@";
+	public const string GETTEXT_PACKAGE = "@GETTEXT_PACKAGE@";
```
[the remark this answers is not in the document]
Comment (mmstick) [reply src/Config.vala.in]: Likely an editor that was configured to use tabs by default, but overrode that setting for `vala` files, but `vala.in` files aren't detected as vala. That would be my guess.

File: src/Utils.vala [hunk 9]
```diff
@@ -81,16 +77,7 @@ namespace Utils {
...
-
-            var seat = Utils.get_seat_instance ();
-            if (seat != null) {
-                try {
-                    seat.switch_to_guest ("");
-                } catch (GLib.Error e) {
-                    stderr.printf ("DisplayManager.Seat error: %s\n", e.message);
-                }
-            }
+            Installer.App.get_instance ().quit ();
```
Comment (cassidyjames) [inline src/Utils.vala]: This is pop-specific. Upstream should still switch to the guest account.

File: src/Application.vala [hunk 10]
```diff
@@ -26,11 +26,20 @@ public class Installer.App : Gtk.Application {
 
     public static bool test_mode;
 
+    private static Installer.App instance = null;
```
Comment (cassidyjames) [inline src/Application.vala]: This and 39–42 are only needed for the pop-specific quit https://github.com/elementary/installer/pull/335/files#r195868345

File: src/Views/AbstractInstallerView.vala [hunk 11]
```diff
@@ -19,6 +19,7 @@
 public abstract class AbstractInstallerView : Gtk.Grid {
     public bool cancellable { get; construct; }
     public unowned Gtk.Widget? previous_view { get; set; }
+    public Gtk.Label? test_label = null;
```
Comment (cassidyjames) [inline src/Views/AbstractInstallerView.vala]: Not needed anymore since we're not setting a timer. Can just be defined in the construct block like before.

File: src/Views/AbstractInstallerView.vala [hunk 12]
```diff
@@ -36,8 +37,12 @@ public abstract class AbstractInstallerView : Gtk.Grid {
         content_area = new Gtk.Grid ();
         content_area.column_spacing = 12;
         content_area.row_spacing = 12;
+        content_area.margin = 12;
```
Comment (cassidyjames) [inline src/Views/AbstractInstallerView.vala]: I _think_ this and 43–45 were pop-specific, but I'm not 100% sure.

File: src/Views/DiskView.vala [hunk 13]
```diff
@@ -48,8 +48,9 @@ public class Installer.DiskView : AbstractInstallerView {
         var install_image = new Gtk.Image.from_icon_name ("system-os-installer", Gtk.IconSize.DIALOG);
         install_image.valign = Gtk.Align.START;
 
-        var install_label = new Gtk.Label (_("Select a Drive"));
-        install_label.hexpand = true;
+        var install_label = new Gtk.Label (_("Select a drive"));
```
Comment (cassidyjames) [inline src/Views/DiskView.vala]: Should be title case like before

File: src/Views/EncryptView.vala [hunk 14]
```diff
@@ -48,21 +48,21 @@ public class EncryptView : AbstractInstallerView {
...
-        var performance_image = new Gtk.Image.from_icon_name ("emoji-objects-symbolic", Gtk.IconSize.LARGE_TOOLBAR);
+        var performance_image = new Gtk.Image.from_icon_name ("utilities-system-monitor-symbolic", Gtk.IconSize.LARGE_TOOLBAR);
 
         var performance_label = new Gtk.Label (_("Drive encryption may minimally impact read and write speed when performing intense tasks."));
         performance_label.max_width_chars = 52;
         performance_label.wrap = true;
         performance_label.xalign = 0;
 
-        var restart_image = new Gtk.Image.from_icon_name ("rotation-allowed-symbolic", Gtk.IconSize.LARGE_TOOLBAR);
+        var restart_image = new Gtk.Image.from_icon_name ("system-restart-symbolic", Gtk.IconSize.LARGE_TOOLBAR);
```
Comment (cassidyjames) [inline src/Views/EncryptView.vala]: @danrabbit we'll either need to make a `system-reboot-symbolic` and change this to that, or revert this icon for now.

File: src/Views/ProgressView.vala [hunk 15]
```diff
@@ -132,204 +156,71 @@ public class ProgressView : AbstractInstallerView {
         installer.on_status (installation_status_callback);
 
         var config = Distinst.Config ();
-        config.flags = Distinst.MODIFY_BOOT_ORDER;
-        config.hostname = "todo";
-        config.lang = "en_US.UTF-8";
-        config.remove = Build.MANIFEST_REMOVE_PATH;
-        config.squashfs = Build.SQUASHFS_PATH;
+        config.flags = Distinst.MODIFY_BOOT_ORDER | Distinst.INSTALL_HARDWARE_SUPPORT;
+        config.hostname = "pop-os";
```
Comment (cassidyjames) [inline src/Views/ProgressView.vala]: Whoops! :smile: We should do something better here.

File: src/Views/SuccessView.vala [hunk 16]
```diff
@@ -62,12 +102,6 @@ public class SuccessView : AbstractInstallerView {
         action_area.add (shutdown_button);
         action_area.add (restart_button);
 
-        Timeout.add_seconds (RESTART_TIMEOUT, () => {
```
Comment (cassidyjames) [inline src/Views/SuccessView.vala]: I _think_ this should remain upstream, as we do want the countdown for automatically restarting. It was removed downstream for Pop!

File: src/Views/TryInstallView.vala [hunk 17]
```diff
@@ -96,8 +96,6 @@ public class Installer.TryInstallView : AbstractInstallerView {
         action_area.add (shutdown_button);
         action_area.add (back_button);
         action_area.add (next_button);
-        action_area.set_child_secondary (shutdown_button, true);
```
Comment (cassidyjames) [inline src/Views/TryInstallView.vala]: I think this should still be a secondary button?

File: src/Widgets/DiskBar.vala [hunk 18]
```diff
@@ -24,104 +24,79 @@ public class Installer.DiskBar: Gtk.Grid {
...
-        attach (bar, 1, 1);
-
-        show_all ();
     }
 
     private void add_legend (string ppath, uint64 size, string fs, string? vg, Gtk.Popover? menu) {
         var fill_round = new FillRound ();
         fill_round.width_request = fill_round.height_request = 14;
-        fill_round.valign = Gtk.Align.CENTER;
+        fill_round.set_valign (Gtk.Align.CENTER);
```
Comment (cassidyjames) [inline src/Widgets/DiskBar.vala]: Another setter vs prop

**Comment by cassidyjames** [comment]
@mmstick a minimum-viable PR to get things building again with the latest distinst changes would be super useful, then other PRs can be merged as they're ready. It looks like the main issue I'm seeing in master is missing arguments for `Distinst.Disc.contains_mount ()`

**Comment by mmstick** [comment]
@cassidyjames I think I've addressed all your prior comments. Will work on getting things split up now.

**Comment by cassidyjames** [comment]
@mmstick awesome. I know it's a lot, but I want to make sure this is a clean of a merge as possible for other contributors, and I don't want to get bits held up due to unrelated changes. I appreciate you working through it!

File: src/Application.vala [hunk 19]
```diff
@@ -31,6 +31,7 @@ public class Installer.App : Gtk.Application {
         flags = ApplicationFlags.FLAGS_NONE;
         Intl.setlocale (LocaleCategory.ALL, "");
         add_main_option_entries (INSTALLER_OPTIONS);
+        instance = this;
```
Comment (cassidyjames) [inline src/Application.vala]: I think this is still leftover from the pop-specific quit for demo mode thing?

File: src/MainWindow.vala [hunk 20]
```diff
@@ -40,7 +40,7 @@ public class Installer.MainWindow : Gtk.Dialog {
             deletable: false,
             height_request: 700,
             icon_name: "system-os-installer",
-            resizable: false,
+            resizable: true,
```
Comment (cassidyjames) [inline src/MainWindow.vala]: I remember this fixed a specific bug, but I don't recall if it was pop-specific due to the artwork or not. @danrabbit thoughts on making the installer resizable?

Comment (mmstick) [reply src/MainWindow.vala]: The issue was for disks that had many partitions -- the window is too small to comfortably accommodate disks with more than a few partitions.

**Comment by mmstick** [comment]
I think I've gotten all the important parts split apart now. What remains are changes to implement the InstallOptions feature of distinst, which will be used to enable installing alongside another OS, and many other installation types.

File: src/Views/ErrorView.vala [hunk 21]
```diff
@@ -61,6 +61,8 @@ public class ErrorView : AbstractInstallerView {
 
         var grid = new Gtk.Grid ();
         grid.row_spacing = 6;
+        grid.margin_left = 24;
+        grid.valign = Gtk.Align.CENTER;
```
Comment (cassidyjames) [inline src/Views/ErrorView.vala]: `margin_left` is deprecated and not RTL friendly; use `margin_start` instead. Also, I think these are pop-specific styling?

File: src/Views/ProgressView.vala [hunk 22]
```diff
@@ -342,8 +212,7 @@ public class ProgressView : AbstractInstallerView {
                     var new_disk = new Distinst.Disk (m.parent_disk);
                     if (new_disk == null) {
                         warning ("could not find physical device: '%s'\n", m.parent_disk);
```
Comment (cassidyjames) [inline src/Views/ProgressView.vala]: Should this be a `critical ()` like the others?

File: src/Views/ProgressView.vala [hunk 23]
```diff
@@ -398,20 +265,20 @@ public class ProgressView : AbstractInstallerView {
         }
 
         foreach (Installer.Mount m in lvm_devices) {
-            var vg = m.parent_disk.offset (12);
+            var vg = m.parent_disk.offset (12).replace ("--", "-");
```
Comment (cassidyjames) [inline src/Views/ProgressView.vala]: Why is this needed?

Comment (mmstick) [reply src/Views/ProgressView.vala]: /dev/mapper has special rules for handling the `-` character in the device path, because the `-` character is used to separate the volume group identifier from the logical volume identifier. Someone had the installer fail because they were trying to install to a VG with `-` in the vg identifier.

File: src/Views/ProgressView.vala (hunk 22 above)
Comment (mmstick) [reply src/Views/ProgressView.vala]: Yes, it should.

File: src/Views/ProgressView.vala (hunk 23 above)
Comment (cassidyjames) [reply src/Views/ProgressView.vala]: It might be worth adding a comment for that, then; otherwise this is spooky magic at a distance

File: src/Views/CheckView.vala [hunk 24]
```diff
@@ -252,7 +254,7 @@ public class Installer.CheckView : AbstractInstallerView {
         var title_label = new Gtk.Label (title);
         title_label.get_style_context ().add_class ("h2");
         title_label.wrap = true;
-        title_label.max_width_chars = 60;
+        title_label.max_width_chars = 30;
```
Comment (cassidyjames) [inline src/Views/CheckView.vala]: Do we know why this changed?

Comment (mmstick) [reply src/Views/CheckView.vala]: Pop seems to use 30 for everything, but elementary is using 60.

File: src/Views/DiskView.vala [hunk 25]
```diff
@@ -108,64 +109,63 @@ public class Installer.DiskView : AbstractInstallerView {
...
+
                         next_button.sensitive = true;
                     } else {
                         next_button.sensitive = false;
                     }
                 });
 
                 enabled_buttons += disk_button;
+            } else {
+                disk_button.set_sensitive (false);
```
Comment (cassidyjames) [inline src/Views/DiskView.vala]: Should probably be `.sensitive = false` like the others

Comment (mmstick) [reply src/Views/DiskView.vala]: Seems like `vala-lint` should check for these.

File: src/Views/ProgressView.vala [hunk 26]
```diff
@@ -112,6 +112,27 @@ public class ProgressView : AbstractInstallerView {
...
+     * distinst as its original name, though.
+     *
+     * Note: the device path specified here should refer to a LVM or LUKS device, rather
+     * than a logical volume that exists on a LVM device.
+     */
+    private string extract_volume_group_from (string device_path) {
+        return device_path
+            // Remove the `/dev/mapper` prefix
+            .offset (12)
+            // Replace all occurrences of `--` with `-`
```
Comment (cassidyjames) [inline src/Views/ProgressView.vala]: This still just says the what (which is obvious from the line below), and not the why. The description about the special treatment of `-` would make sense in this comment.

Comment (mmstick) [reply src/Views/ProgressView.vala]: The why is explained in the function's description.

Comment (cassidyjames) [reply src/Views/ProgressView.vala]: Ah, I see that now. 👍

File: src/Views/TryInstallView.vala [hunk 27]
```diff
@@ -155,7 +155,22 @@ public class Installer.TryInstallView : AbstractInstallerView {
             }
         });
 
+        custom_button.clicked.connect (() => {
```
Comment (cassidyjames) [inline src/Views/TryInstallView.vala]: This looks like a copy-paste of the block above. Am I missing something?

Comment (mmstick) [reply src/Views/TryInstallView.vala]: @cassidyjames Feel free to do whatever you want with these PRs. The diff between Pop!_OS's installer and elementary's installer is so large now, even from the time of submitting these PRs, that there isn't much point in me spending time on them, when it'd be easier for elementary to rebase themselves on Pop's repo and make whatever changes they want instead.

**Comment by danrabbit** [comment]
Closing since there's been no movement it sounds like there is no an intention for there to be
